crates/core/app/tests/common/test_node_builder_ext.rs

```diff
@@ -0,0 +1,120 @@
+use {
+    penumbra_genesis::AppState,
+    penumbra_mock_consensus::{builder::Builder, keyring::Keys},
+    penumbra_proto::{
+        core::keys::v1::{GovernanceKey, IdentityKey},
+        penumbra::core::component::stake::v1::Validator as PenumbraValidator,
+    },
+    tap::Tap,
+};
+
+/// Penumbra-specific extensions to the mock consensus builder.
+pub trait BuilderExt: Sized {
+    /// The error thrown by [`with_penumbra_auto_app_state`]
+    type Error;
+    /// Add the provided Penumbra [`AppState`] to the builder.
+    ///
+    /// This will inject any configured validators into the state before serializing it into bytes.
+    fn with_penumbra_auto_app_state(self, app_state: AppState) -> Result<Self, Self::Error>;
+}
+
+impl BuilderExt for Builder {
+    type Error = anyhow::Error;
+    fn with_penumbra_auto_app_state(self, app_state: AppState) -> Result<Self, Self::Error> {
+        // Generate a penumbra validator using the test node's consensus keys (if they exist).
+        // Eventually, we may wish to generate and inject additional definitions, but only a single
+        // validator is supported for now.
+        let app_state = match self
+            .keys
+            .as_ref()
+            .map(generate_penumbra_validator)
+            .inspect(log_validator)
+            .map(std::iter::once)
+        {
+            Some(validator) => app_state_with_validators(app_state, validator)?,
+            None => app_state,
+        };
+
+        // Serialize the app state into bytes, and add it to the builder.
+        serde_json::to_vec(&app_state)
+            .map_err(Self::Error::from)
+            .map(|s| self.app_state(s))
+    }
+}
```

Comment:
:point_up: here's the meat of it

Reply:
It's a little hard for me to reason about what the code is doing. We want to:

1. Generate a list of penumbra validator definitions corresponding to the consensus keys specified in the builder. There must be at least one.

2. Inject those validator definitions into the provided AppState

3. Create allocations of delegation tokens for each validator, controlled by some address. Otherwise the validator won't have any voting power and will be immediately deleted.

Could we make it more explicit?

Reply:
e66628697 among other things, rewrites this function. i think now that should more clearly reflect steps (1) and (2)!